### PR TITLE
Add DateTime64 data type.

### PR DESCRIPTION
**Pull request by Enmk**
Based on #38 
New Column type: `ColumnDateTime64`
Sligtly updated some underlying code
Added unit and integrration tests for ColumnDateTime64

Closes #38

### CODE REVIEW

File: clickhouse/columns/decimal.cpp [hunk 1]
```diff
@@ -83,12 +83,16 @@ void ColumnDecimal::Append(const std::string& value) {
...
-        return data_->As<ColumnInt128>()->At(i);
+    switch (data_->Type()->GetCode()) {
+        case Type::Int32:
+            return static_cast<Int128>(data_->As<ColumnInt32>()->At(i));
+        case Type::Int64:
+            return static_cast<Int128>(data_->As<ColumnInt64>()->At(i));
+        case Type::Int128:
+            return data_->As<ColumnInt128>()->At(i);
+        default:
+            assert(false && "Invalid data_ column type in ColumnDecimal");
```
Comment (traceon) [inline clickhouse/columns/decimal.cpp:94]: Please, do not use `assert()` for failures that may happen during normal operation (or depend on input data) and not only in testing (which means pretty much never.) These asserts will be no-op in release builds, leaving the code without that protection.

File: clickhouse/columns/numeric.cpp [hunk 2]
```diff
@@ -10,9 +10,9 @@ ColumnVector<T>::ColumnVector()
 }
 
 template <typename T>
-ColumnVector<T>::ColumnVector(const std::vector<T>& data)
+ColumnVector<T>::ColumnVector(std::vector<T> data)
     : Column(Type::CreateSimple<T>())
-    , data_(data)
+    , data_(std::move(data))
```
[the remark this answers is not in the document]
Comment (traceon) [reply clickhouse/columns/numeric.cpp:15]: If there are cases, when the passed argument is a temporary that could be moved-in, then better to add a move c-tor: `ColumnVector(std::vector<T>&& data)`

File: ut/client_ut.cpp [hunk 3]
```diff
@@ -631,6 +658,71 @@ TEST_P(ClientCase, Decimal) {
...
+    client_->Select("SELECT dt FROM test_clickhouse_cpp.datetime64",
+        [&total_rows, &data](const Block& block) {
+            total_rows += block.GetRowCount();
+            if (block.GetRowCount() == 0) {
+                return;
+            }
+
+            ASSERT_EQ(1U, block.GetColumnCount());
+            if (auto col = block[0]->As<ColumnDateTime64>()) {
+                ASSERT_EQ(data.size(), col->Size());
```
Comment (traceon) [inline ut/client_ut.cpp:715]: Does this assume, that you always read an entire `data`-sized block here?

File: ut/columns_ut.cpp [hunk 4]
```diff
@@ -114,6 +152,135 @@ TEST(ColumnsCase, DateAppend) {
...
+    ASSERT_EQ(0u, column->Size());
+    ASSERT_EQ(data.size(), column2->Size());
+
+    // Clearing doesn't affects precision value.
+    ASSERT_EQ(6ul, column->GetPrecision());
+
+    for (size_t i = 0; i < data.size(); ++i) {
+        ASSERT_EQ(data[i], column2->At(i));
+    }
+}
```
Comment (traceon) [inline ut/columns_ut.cpp:224]: This checks only that one direction is OK after the swap, right? One typical bug of a `swap` implementation, is when after a swap the destinations contains what was in the source, but the source is left empty.

File: ut/columns_ut.cpp [hunk 5]
```diff
@@ -114,6 +152,135 @@ TEST(ColumnsCase, DateAppend) {
...
+    }
+
+    ASSERT_EQ(data.size(), column->Size());
+    for (size_t i = 0; i < data.size(); ++i) {
+        ASSERT_EQ(data[i], column->At(i));
+    }
+}
+
+TEST(ColumnsCase, DateTime64_Clear) {
+    auto column = std::make_shared<ColumnDateTime64>(6ul);
```
Comment (traceon) [inline ut/columns_ut.cpp:188]: It is worth adding `ASSERT_EQ(0u, column->Size());` to ensure that a column is created empty.

File: ut/columns_ut.cpp [hunk 6]
```diff
@@ -114,6 +152,135 @@ TEST(ColumnsCase, DateAppend) {
...
+        auto slice = column->Slice(offset, count)->As<ColumnDateTime64>();
+
+        ASSERT_EQ(count, slice->Size());
+        ASSERT_EQ(column->GetPrecision(), slice->GetPrecision());
+
+        for (size_t i = offset; i < offset + count; ++i) {
+            ASSERT_EQ(data[i], slice->At(i - offset));
+        }
+    }
+}
```
Comment (traceon) [inline ut/columns_ut.cpp:275]: What happens on out-of-bound slices (5 cases)?

**Review by traceon — CHANGES_REQUESTED** [review]
Left comments.

File: clickhouse/types/type_parser.cpp [hunk 7]
```diff
@@ -9,6 +9,7 @@
 namespace clickhouse {
 
 bool TypeAst::operator==(const TypeAst & other) const {
+    // TODO: operator == must be a friend function.
```
[the remark this answers is not in the document]
Comment (Enmk) [reply clickhouse/types/type_parser.cpp:12]: That is a leftovers from previous PR from another author, and I guess we will never figure out why... Let me remove it

File: ut/client_ut.cpp [hunk 8]
```diff
@@ -712,7 +712,6 @@ TEST_P(ClientCase, DateTime64) {
 
             ASSERT_EQ(1U, block.GetColumnCount());
             if (auto col = block[0]->As<ColumnDateTime64>()) {
-                ASSERT_EQ(data.size(), col->Size());
                 for (size_t i = 0; i < col->Size(); ++i) {
                     EXPECT_EQ(data[i], col->At(i)) << " at index: " << i;
```
Comment (traceon) [inline ut/client_ut.cpp:716]: Are the indices absolute? Or they are offsets within the current block?
If the latter, then the code still relies on the assumption that the entire data will be received in the [first] block. Maybe, technically it is what will always happen here, but if the data is set to size of, e.g., `100 000 000`, will this code still be correct?

File: clickhouse/columns/decimal.cpp [hunk 9]
```diff
@@ -91,7 +90,7 @@ Int128 ColumnDecimal::At(size_t i) const {
         case Type::Int128:
             return data_->As<ColumnInt128>()->At(i);
         default:
-            assert(false && "Invalid data_ column type in ColumnDecimal");
+            throw std::runtime_error("Invalid data_ column type in ColumnDecimal");
             return 0;
```
Comment (traceon) [inline clickhouse/columns/decimal.cpp:94]: Extra return, looks like.

File: ut/columns_ut.cpp [hunk 10]
```diff
@@ -274,6 +274,23 @@ TEST(ColumnsCase, DateTime64_Slice) {
...
+    // Non-Empty slice on empty column
+    EXPECT_EQ(0u, column->Slice(0, 10)->Size());
+
+    const auto data = MakeDateTime64s();
+    for (const auto & v : data) {
+        column->Append(v);
+    }
+
+    EXPECT_EQ(column->Slice(0, data.size() + 1)->Size(), data.size());
+    EXPECT_EQ(column->Slice(data.size() + 1, 1)->Size(), 0u);
```
Comment (traceon) [inline ut/columns_ut.cpp:291]: Maybe, also this?

```c++
     EXPECT_EQ(column->Slice(data.size() / 2, data.size() / 2 + 2)->Size(), data.size()); 
```

**Review by traceon — COMMENTED** [review]
Some comments.

File: ut/columns_ut.cpp (hunk 4 above)
Comment (Enmk) [reply ut/columns_ut.cpp]: fixed

File: ut/columns_ut.cpp (hunk 5 above)
Comment (Enmk) [reply ut/columns_ut.cpp:188]: We check that in other tests.

File: ut/columns_ut.cpp (hunk 6 above)
Comment (Enmk) [reply ut/columns_ut.cpp:275]: Added tests

File: ut/client_ut.cpp (hunk 8 above)
Comment (Enmk) [reply ut/client_ut.cpp:716]: Fixed

File: ut/client_ut.cpp (hunk 3 above)
Comment (Enmk) [reply ut/client_ut.cpp]: good note, fixed

File: clickhouse/columns/decimal.cpp (hunk 1 above)
Comment (Enmk) [reply clickhouse/columns/decimal.cpp]: Ok

File: clickhouse/columns/numeric.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (Enmk) [reply clickhouse/columns/numeric.cpp]: Ok, no problem, I guess it is time for me to re-read core guidelines.

File: clickhouse/columns/numeric.cpp [hunk 11]
```diff
@@ -10,7 +10,14 @@ ColumnVector<T>::ColumnVector()
...
+ColumnVector<T>::ColumnVector(const std::vector<T> & data)
+    : Column(Type::CreateSimple<T>())
+    , data_(data)
+{
+}
+
+template <typename T>
+ColumnVector<T>::ColumnVector(std::vector<T> && data)
     : Column(Type::CreateSimple<T>())
     , data_(data)
```
Comment (traceon) [inline clickhouse/columns/numeric.cpp:22]: Then, this one should be passed further as an rvalue here: `, data_(std::move(data))`

File: ut/client_ut.cpp (hunk 8 above)
Comment (traceon) [reply ut/client_ut.cpp:716]: I mean, it it correct to use `i` that iterates over `col` ans an index for `data`?

**Review by traceon — COMMENTED** [review]
One more.

File: ut/performance_tests.cpp [hunk 12]
```diff
@@ -80,7 +80,16 @@ class ColumnPerformanceTest : public ::testing::Test {
...
 
+// Turns out this is the easiest way to skip test with current version of gtest
+#ifndef NDEBUG
+#  define SKIP_IN_DEBUG_BUILDS() do { std::cerr << "Test skipped...\n"; return; } while(0)
+#else
+#  define SKIP_IN_DEBUG_BUILDS() (void)(0)
+#endif
+
 TYPED_TEST_P(ColumnPerformanceTest, SaveAndLoad) {
+    SKIP_IN_DEBUG_BUILDS();
```
Comment (traceon) [inline ut/performance_tests.cpp:91]: FYI, Google Test has `GTEST_SKIP()` starting from some version, and, alternatively, you can prepend `DISABLED_` to the name of a test to disable it (should be available in the bundled `gtest` repo). Not an issue here, though.

**Review by traceon — APPROVED** [review]

**Pull request merged**